### PR TITLE
consensus/pbft: abstract the proposer selection

### PR DESCRIPTION
**Pull request by markya0616**

### CODE REVIEW

File: consensus/pbft/validator.go [hunk 1]
```diff
@@ -16,90 +16,28 @@
...
+	Size() int
+	// Return the validator array
+	List() []Validator
+	// Get validator by index
+	GetByIndex(i uint64) Validator
+	// Get validator by address
+	GetByAddress(addr common.Address) Validator
+	// Get current proposer
+	GetProposer() Validator
+	// Check wether the validator with address is a proposer
```
Comment (bailantaotao) [inline consensus/pbft/validator.go]: whether

File: consensus/pbft/backends/simple/engine.go [hunk 2]
```diff
@@ -237,14 +242,16 @@ func (sb *simpleBackend) Stop() error {
...
-func (sb *simpleBackend) initValidatorSet(chain consensus.ChainReader) {
+func (sb *simpleBackend) initValidatorSet(chain consensus.ChainReader) bool {
 	currentHeader := chain.CurrentHeader()
-	addrs := getValidatorSet(currentHeader)
-	vals := make([]*pbft.Validator, len(addrs))
-	for i, addr := range addrs {
-		vals[i] = pbft.NewValidator(addr)
+	// get the validator byte array and feed into validator set
+	length := len(currentHeader.Extra)
+	valSet, r := validator.NewSet(currentHeader.Extra[extraVanity : length-extraSeal])
```
Comment (tailingchen) [inline consensus/pbft/backends/simple/engine.go]: It's better to pass whole extra data to `NewSet`. `extraVanity` and `extraSeal`  are the parts of logic to parse validators.

File: consensus/pbft/backends/simple/engine.go [hunk 3]
```diff
@@ -237,14 +242,16 @@ func (sb *simpleBackend) Stop() error {
...
+	if !r || valSet == nil {
+		return false
 	}
-	sb.valSet = pbft.NewValidatorSet(vals)
+	sb.valSet = valSet
+	return true
 }
 
 func getValidatorSet(block *types.Header) []common.Address {
```
Comment (tailingchen) [inline consensus/pbft/backends/simple/engine.go]: remove it if redundant

File: consensus/pbft/backends/simple/engine.go (hunk 2 above)
Comment (markya0616) [inline consensus/pbft/backends/simple/engine.go]: 我覺得放外面比較好耶   只有中間的byte array跟validator set有關   前面跟後面的東西是跟consensus有關的

File: consensus/pbft/backends/simple/engine.go (hunk 3 above)
Comment (markya0616) [inline consensus/pbft/backends/simple/engine.go]: OK

File: consensus/pbft/backends/simple/engine.go [hunk 4]
```diff
@@ -227,7 +230,9 @@ func (sb *simpleBackend) HandleMsg(peerID string, data []byte) error {
 
 // Start implements consensus.PBFT.Start
 func (sb *simpleBackend) Start(chain consensus.ChainReader) error {
-	sb.initValidatorSet(chain)
+	if !sb.initValidatorSet(chain) {
```
Comment (markya0616) [inline consensus/pbft/backends/simple/engine.go]: OK

**Pull request merged**